### ISSUE TITLE
TLS Websockets

### ISSUE THREAD

**Issue by mishoboss** [mcollina/mosca#298]
Hello, 
Is there a way to have secured websockets (wss) in Mosca? I see there is a "secure" config object, but it affects only the raw TCP, not the websockets. The "http" config object, that is for websockets, don't have any secure options.

I'm pretty new to MQTT and Mosca, so please excuse me if I ask a stupid question or I do it in wrong place (if there is a better place, please point it).

**Comment by PritamUpadhyay** [comment]
Can you show your server file?

**Comment by mishoboss** [comment]
var currentDir = __dirname.split("/").pop();

var PORT=8001;
var SECURE_PORT=8002;
var HTTP_PORT=8003;



var mosca = require('mosca');
var SECURE_KEY = __dirname + '/../key.pem';
var SECURE_CERT = __dirname + '/../cert.pem';

var moscaSettings = {
  port: PORT,
  stats: false,
  //backend: ascoltatore,
  // persistence: {
  //   factory: mosca.persistence.Mongo,
  //   url: 'mongodb://localhost:27017/mqtt'
  // },
  secure : {
    keyPath: SECURE_KEY,
    certPath: SECURE_CERT,
    port: SECURE_PORT
  },

  allowNonSecure: true,

  http: {
    port: HTTP_PORT,
    bundle: true,
    static: './'
  }
};

var server = new mosca.Server(moscaSettings);
server.on('ready', setup);

server.on('clientConnected', function(client) {
    console.log('client connected', client.id);
});

// fired when a message is received
server.on('published', function(packet, client) {
  if(packet.topic == '/user/login'){
    
  }
  console.log('Published', packet.payload);
});

// fired when the mqtt server is ready
function setup() {
  console.log('Mosca server is up and running')
}

**Comment by mcollina** [comment]
just pass in an https property.

**Comment by mishoboss** [comment]
What about keys? Do I have to put some pem paths like in "secure" object?

**Comment by mcollina** [comment]
Same as secure, but within the https object.

**Comment by mishoboss** [comment]
Seems not working.

https: {
    keyPath: SECURE_KEY,
    certPath: SECURE_CERT,
    port: HTTP_PORT,
    bundle: true,
    static: './'
  }



This is how I try to connect:
`var client = mqtt.connect('ws://{IP}:9003');`
It shows a pending ws connection, but mqtt doesn't work above.

If I try `var client = mqtt.connect('wss://{IP}:9003');` it can't connect at all.

**Comment by mcollina** [comment]
Sorry, you should drop the keyPath and certPath in the secure field:

See: https://github.com/mcollina/mosca/blob/master/test/server_websocket_secure.js#L19-L27

**Comment by mishoboss** [comment]
Nothing changes. I had the secure filed before too. 


This is the client:

```js
var client = mqtt.connect('ws://{IP}:9003'); // you add a ws:// url here
    client.subscribe("mqtt/demo");

    client.on("message", function(topic, payload) {
      alert([topic, payload].join(": "));
      client.end();
    });

    client.on("connect", function() {
      alert('CONNECTED!');
      client.publish("mqtt/demo", "hello world!");
    });
```


This is the server:

```js
var currentDir = __dirname.split("/").pop();

  PORT=9001;
  SECURE_PORT=9002;
  HTTP_PORT=9003;

var mosca = require('mosca');
var SECURE_KEY = __dirname + '/../key.pem';
var SECURE_CERT = __dirname + '/../cert.pem';


var moscaSettings = {
  port: PORT,
  stats: false,
 
  secure : {
    keyPath: SECURE_KEY,
    certPath: SECURE_CERT
    //port: SECURE_PORT
  },

  //allowNonSecure: true,

  https: {
    port: HTTP_PORT,
    bundle: true,
    static: './'
  }
};

var server = new mosca.Server(moscaSettings);
server.on('ready', setup);

server.on('clientConnected', function(client) {
    console.log('client connected', client.id);
});

// fired when a message is received
server.on('published', function(packet, client) {
  console.log('Published', packet.payload);
});

// fired when the mqtt server is ready
function setup() {
  console.log('Mosca server is up and running')
}
```

**Comment by mcollina** [comment]
it should work with the wss protocol.

**Comment by mishoboss** [comment]
With wss I get this:

`WebSocket connection to 'wss:/{IP}9003/' failed: WebSocket opening handshake was canceled`

**Comment by mcollina** [comment]
I think you are missing `rejectUnauthorized: false` in your connect() call

**Comment by mcollina** [comment]
Are your certificates self-signed or not?

**Comment by mishoboss** [comment]
Yes, it's self-signed. Added "rejectUnauthorized: false" as option to connect(), but still the same. Do I have to attach the .pem files in the client too? BTW, I use the mqtt.js in a browser via Browsify (if this matters at all). Is there a good and full example of TLS Websockets with Mosca and MQTT.js?

**Comment by mcollina** [comment]
That _might_ matter. Can you please try the same snippet from node?

In the browser the problem is that you have to accept the self-signed certificate.

**Comment by mishoboss** [comment]
Can't connect. I don't get any console.logs.

```js
var mqtt    = require('mqtt');
var client = mqtt.connect('wss://{IP}:9003', {rejectUnauthorized: false}); // you add a ws:// url here
client.subscribe("mqtt/demo");

client.on("message", function(topic, payload) {
  console.log([topic, payload].join(": "));
  client.end();
});

client.on("connect", function() {
  console.log('CONNECTED!');
  client.publish("mqtt/demo", "hello world!");
});
```

**Comment by mcollina** [comment]
You have the wrong port in the client, in the broker example is 8003 while in the client is 9003.

**Comment by mcollina** [comment]
that same code (with that change) works here

**Comment by mishoboss** [comment]
No, it's not that. I have two versions - one using 8003 and one using 9003. I copied code from both, that's why. See my second code post with full server and client code.

**Comment by mishoboss** [comment]
OK, here is what I found next. I saw that when script is run two ports are opened - 9003 and 3001. I googled the second port and it came out that it's a default port for express. Having a "port" param in the "https" object overwrites the default express port.

```js
secure : {
    keyPath: SECURE_KEY,
    certPath: SECURE_CERT,
    port: SECURE_PORT
  },

  https: {
    port: HTTP_PORT,
    bundle: true,
    static: HTTP_PATH
  }
```

Now I have the two correct ports listening in the server - one is for raw TCP and the other is for Websockets. But that didn't changed anything. The client still couldn't connect via Websockets. However I run the client script from node, not as Browserify script in browser and it connected successfully. So, now the problem seems something with the "browsification" of mqtt.js. This is what I get in console:

`WebSocket connection to 'wss://{IP}:9003/' failed: WebSocket opening handshake was canceled`
